python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+        pass
+
+    def start_workflow(
+        workflow: str,
+        *args,
+        id: str = None,
+        task_queue: str = None,
+        retry_policy: RetryPolicy = None,
+        # Note, we are embedding signal-with-start in here
+        start_signal: str = None,
+        start_signal_args=[],
+        # ... additional options omitted for brevity
+    ) -> WorkflowHandle[Any]:
+        pass
+
+    # In a future phase:
+    # @overload
+    # def start_workflow(workflow: Callable[P, T], *args: P.args, ...) -> WorkflowExecution[T]
+
+    # Does not validate whether a workflow exists
+    def workflow(
+        workflow_id: str,
+        run_id: str = None
+        # ...
+    ) -> WorkflowHandle[Any]:
+        pass
+
+
+class WorkflowHandle(Generic[T]):
+    def __init__(
+        self,
+        client: Client,
+        workflow_id: str,
+        *,
+        run_id: str = None
+        # ...
+    ) -> None:
+        pass
+
+    def __await__(self) -> T:
+        pass
+
+    async def cancel(self):
+        pass
+
+    async def describe(self):
+        pass
+
+    async def query(self, name: str, *args):
+        pass
+
+    async def signal(self, name: str, *args):
+        pass
+
+    async def terminate(self, *, reason: str = None):
+        pass
+```
+
+Notes:
+
+* We are intentionally defining a `Client` instead of a `WorkflowClient` since this will be general purpose
+  * Having a general purpose client keeps things simple, but at least for now I can't see defining any more than the two
+    methods and the one property
+  * We could make an abstract base class if we want for typing/purposes that just has the `service` property
+    unimplemented
+* We call it `start_workflow` instead of just `start` to be clearer
+* We have embedded `start_signal` and `start_signal_args` as parameters, so if the former is present we will perform a
+  signal with start. Is this ok?
+  * This is an intentional deviation from other SDKs, but makes logical sense
+* The client supports a default task queue, do we want this?
+
+### Worker
+
+Exported from `temporalio` (but probably in `temporalio/worker/worker.py`, exported from `temporalio.worker` then
+exported from `temporalio`):
+
+```python
+class Worker:
+    def __init__(
+        self,
+        client: temporalio.Client,
+        *,
+        # Uses default in client if not present (or errors if no default)
+        task_queue: str = None
+        # ...
+    ) -> None:
+        pass
+
+    # Cannot be called after run started. Typings and separate @overload
+    # signatures omitted for brevity.
+    def register_activity(
+        # This must be a callable. Can be a lambda.
+        activity,
+        *,
+        # Set the name override. If not set, uses the @activity.run decorator
+        # name. If that's not set, uses the function/method name or class name
+        # if it has __call__. If it's a lambda or nested function, a name must
```

Comment:
I disagree. I think simplicity is important here, even if it means implicitness by default. I don't think there's more value in a dynamic situation like Python making people type more in this case.

I think TS is unnecessarily complex to do this stuff in. I think we should really optimize for simplicity from the caller POV with defaults that can be overridden.

The lambda use case is just so I don't have to care about the type of callable. Simply put "if it's callable, you can register it as an activity (but you might have to provide a name)".

Reply:
It's a super minimal amount of extra typing that makes things more explicit in a way that prevents a certain type of bug we have seen users run into in the real world. That's a good thing, IMO. If all your worker code is always bundled together this is 0% an issue, but if you're calling activities from other languages, you have to understand it's derived from the fn name (to be fair that's pretty obvious) which means you also have to be smart enough to realize you don't get to change that function name once you start relying on it. That's the footgun.

Reply:
It's also a nice signal to other non-temporal code you might have in the same process that "Hey, this is sometimes used as a temporal activity".

Reply:
To confirm, you are suggesting users should always have to provide a string literal for every activity even if, say, they always want it the function name? I think this footgun is only with some, others recognize the function name will be used as the activity name. It's kinda unfair that Go and Java get this simplicity benefit that this newer SDK will not.

Reply:
Yeah. I really don't think it's meaningfully simpler. It's marginally less _verbose_, absolutely, but I just can't see someone getting frustrated because they had to provide a name. Alternatively like in TS / mentioned here: https://github.com/temporalio/proposals/pull/49#discussion_r792249398 if we require the user to provide an explicit mapping I think that's good too.

But, yes, if I had the chance I would undo what Go and Java do where the name is implicitly derived because I've seen it confuse multiple users.

Reply:
I think TS kinda has to because getting the function name in JS at runtime is not always clear/easy w/ how often arrow functions are used and refs passed around. However in Python it seems pretty common to use function name for this stuff.

I have seen it confuse users too, but am not convinced that it deserves to be disallowed because of those users. If we knew how many users liked the inference, I think we could decide better.